.github/workflows/publish-release.yml

```diff
@@ -15,32 +15,60 @@ on:
         default: "main"
 
 jobs:
-  checks:
+  publish:
     runs-on: "ubuntu-latest"
     steps:
       - name: Setup python
         uses: actions/setup-python@v4.5.0
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: "3.11"
 
       - name: Install Hatch
         run: python -m pip install hatch==1.6.3
 
       - uses: actions/checkout@v3.3.0
         with:
           ref: ${{ inputs.repo_release_ref }}
+          fetch-depth: 0 # checkout all history, needed for hatch versioning
 
-      - name: Set SETUPTOOLS_SCM_PRETEND_VERSION
-        run: echo "SETUPTOOLS_SCM_PRETEND_VERSION=$(hatch version | cut -d+ -f1)" >> $GITHUB_ENV
+      # Run hatch version once to avoid additional output ("Setting up build environment for missing dependencies")
+      # during the next step
+      - name: Hatch version
+        run: hatch version
 
-      - name: Test SETUPTOOLS_SCM_PRETEND_VERSION
-        run: echo $SETUPTOOLS_SCM_PRETEND_VERSION
+      # Hatch versioning is based on git (using hatch-vcs). If there is no explicit tag for the commit we're trying to
+      # publish, hatch versioning strings will have this format: 0.19.0.dev52+g9f7dc7d
+      # With the string after '+' being the 'g<short-sha>' of the commit.
+      #
+      # However, PyPI doesn't allow '+' in version numbers (no PEP440 local versions allowed on PyPI).
+      # To work around this, we override the version string by setting the SETUPTOOLS_SCM_PRETEND_VERSION env var
+      # to the version string without the '+' and everything after it.
+      # We then only actual publish such releases on the main branch to guarantee the dev numbering scheme remains
+      # unique.
+      # Note that when a tag *is* present (i.e. v0.19.0), hatch versioning will return the tag name (i.e. 0.19.0)
+      # and this step has no effect, ie. SETUPTOOLS_SCM_PRETEND_VERSION will be the same as `hatch version`.
+      - name: Set SETUPTOOLS_SCM_PRETEND_VERSION
+        run: |
+          echo "SETUPTOOLS_SCM_PRETEND_VERSION=$(hatch version | cut -d+ -f1)" >> $GITHUB_ENV
 
       - name: Build (gitlint-core)
-        run: |
-          hatch build
+        run: hatch build
         working-directory: ./gitlint-core
 
       - name: Build (gitlint)
-        run: |
-          hatch build
+        run: hatch build
```

Comment:
The recommended PyPA tool (`build`) creates a wheel out of an sdist, smoke-testing the behavior of pip and making sure that it's possible to have a working wheel w/o building it from a Git checkout. Have you checked that it's the case for the hatch's build front-end?

Reply:
Merged #463, thanks again for the PR!